accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,375 @@
+
+### Extension Conflict Resolution
+
+If more than one extension applies to a specific member invocation, then if exactly one of them is more specific than all the others, that one is chosen. Otherwise it is a compile-time error.
+
+An extension is more specific than another extension if its `on` type is a subtype of the `on` type of the other extension. If the extension is generic, first do instantiation to bounds, then use the `on` type of that when comparing specificity of extensions.
+
+If an extension's `on` type has a trailing `?`, say `T?`, then it is considered a supertype of the corresponding type `T` without the trailing `?`, as a subtype of any supertype of `T` that also has a trailing `?`, and otherwise unrelated to all other types. So, `on int?` is less specific than `on int`, and `on int?` is more specific than `on num?`. There is no relation between `on num` and `on int?`.
+
+That is, the specificity of an extension is independent of the type it is used at. 
+
+Example:
+
+```dart
+extension SmartIterable<T> on Iterable<T> {
+  void doTheSmartThing(void Function(T) smart) {
+    for (var e in this) smart(e);
+  }
+}
+extension SmartList<T> on List<T> {
+  void doTheSmartThing(void Function(T) smart) {
+    for (int i = 0; i < length; i++) smart(this[i]);
+  }
+}
+...
+  List<int> x = ....;
+  x.doTheSmartThing(print);
+```
+
+Here both the extensions apply, but the `SmartList` extension is more specific than the `SmartIterable` extension because `List<dynamic>` &lt;: `Iterable<dynamic>`.
+
+Example:
+
+```dart
+extension BoxCom<T extends Comparable<T>> on Box<Iterable<T>> { foo() {} }
+extension BoxList<T> on Box<List<T>> { foo() {} }
+extension BoxSpec on Box<List<num>> { foo() {} }
+...
+  List<int> x = ...;
+  x.foo();
+```
+
+Here all three extensions apply. The most specific one is `BoxSpec` which is specialized for lists of numbers. If `BoxSpec` had not been in scope, then neither of `BoxCom` or `BoxList` would be more specific than the other. Their `on` types, when instantiated to bounds, are `Box<Iterable<Comparable<dynamic>>>` and `Box<List<dynamic>>` which are unrelated by subtyping.
+
+This is also why we use the instantiated-to-bounds type for comparison. If we used the actual type, bound to the type variable by the current use-case, we would consider `BoxList<int>` to be more specific than `BoxSpec`, but in practice, the `BoxList<int>` extension would not be able to *use* the integer-ness of the type in its code, so `BoxSpec`, which can specialize its code for lists of numbers, is the most precise match.
+
+### Overriding Access
+
+If two or more extensions apply to the same member access, or if a member of the receiver type prevents using an extension method, or if the extension is shadowed, then it is possible to force an extension member invocation:
+
+```dart
+MyList(object).quickSort();
+```
+
+The syntax looks like a constructor invocation, but it does not create a new object.
```

Comment:
Presumably if `MyList` is generic you should be able to pass the explicit generic arguments as well.

Reply:
Absolutely, or have it inferred like here.